effect/jvm/src/test/scala/scalaz/effect/generator/GenIO.scala

```diff
@@ -0,0 +1,145 @@
+    } yield IO.point[E, A](a)
+
+  /**
+   * Given a generator for `A`, produces a generator for `IO[E, A]` using the `IO.async` constructor.
+   */
+  def genAsyncSuccess[E, A: Arbitrary]: Gen[IO[E, A]] =
+    for {
+      a <- Arbitrary.arbitrary[A]
+    } yield IO.async[E, A](cb => cb(ExitResult.Completed(a)))
+
+  /**
+   * Randomly uses either `genSyncSuccess`, or `genAsyncSuccess` with equal probability.
+   */
+  def genSuccess[E, A: Arbitrary]: Gen[IO[E, A]] = Gen.oneOf(genSyncSuccess[E, A], genAsyncSuccess[E, A])
+
+  /**
+   * Given a generator for `A`, produces a generator for `IO[E, A]` using the `IO.fail` constructor.
+   */
+  def genSyncFailure[E: Arbitrary, A]: Gen[IO[E, A]] =
+    for {
+      error <- Arbitrary.arbitrary[E]
+    } yield IO.fail(error)
+
+  /**
+   * Given a generator for `A`, produces a generator for `IO[E, A]` using the `IO.async` constructor.
+   */
+  def genAsyncFailure[E: Arbitrary, A]: Gen[IO[E, A]] =
+    for {
+      error <- Arbitrary.arbitrary[E]
+    } yield IO.async(cb => cb(ExitResult.Failed(error)))
+
+  /**
+   * Randomly uses either `genSyncFailure`, or `genAsyncFailure` with equal probability.
+   */
+  def genFailure[E: Arbitrary, A]: Gen[IO[E, A]] = Gen.oneOf(genSyncFailure[E, A], genAsyncFailure[E, A])
+
+  def genIO[E: Arbitrary: Cogen, A: Arbitrary: Cogen]: Gen[IO[E, A]] =
+    Gen.frequency(15 -> genSuccess[E, A], 1 -> genFailure[E, A])
+
+  /**
+   * Given a generator for `IO[E, A]`, produces a sized generator for `IO[E, A]` which represents a transformation,
+   * by using some random combination of the methods `map`, `flatMap`, `leftMap`, and any other method that does not change
+   * the success/failure of the value, but may change the value itself.
+   */
+  def genLikeTrans[E: Arbitrary: Cogen, A: Arbitrary: Cogen](implicit gen: Arbitrary[IO[E, A]]): Gen[IO[E, A]] =
+    genTransformations(Gen.oneOf(genOfFlatMaps[E, A], genOfMaps[A], genOfLeftMaps[E]))
+
+  /**
+   * Given a generator for `IO[E, A]`, produces a sized generator for `IO[E, A]` which represents a transformation,
+   * by using methods that can have no effect on the resulting value (e.g. `map(identity)`, `io.race(never)`, `io.par(io2).map(_._1)`).
+   */
+  def genIdentityTrans[E: Arbitrary: Cogen, A: Arbitrary: Cogen](implicit gen: Arbitrary[IO[E, A]]): Gen[IO[E, A]] =
+    genTransformations(
+      Gen.oneOf(genOfIdentityFlatMaps[E, A],
+                genOfIdentityMaps[A],
+                genOfIdentityLeftMaps[E],
+                genOfRace[E, A],
+                genOfParallel[E, A])
+    )
+
+  private def genTransformations[E: Arbitrary: Cogen, A: Arbitrary: Cogen](
+    gens: Gen[FunctionCall]
+  )(implicit gen: Arbitrary[IO[E, A]]): Gen[IO[E, A]] =
+    Gen.sized { size =>
+      for {
+        io        <- gen.arbitrary
+        functions <- Gen.listOfN(size, gens)
+      } yield {
+        functions.foldLeft(io) {
+          case (ioResult, f: FunctionCall.FlatMap[E, A]) =>
```

Comment:
`case (ioResult, FunctionCall.FlatMap(flatmappee))` &c. should save you the type patterns.

Reply:
oh, obviously, nvm :P